pkg/controllers/nodepool/lifecycle/controller.go

```diff
@@ -0,0 +1,137 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lifecycle
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+	nodepoolutil "github.com/aws/karpenter-core/pkg/utils/nodepool"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Controller is hash controller that constructs a hash based on the fields that are considered for static drift.
+// The hash is placed in the metadata for increased observability and should be found on each object.
+type Controller struct {
+	kubeClient client.Client
+}
+
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		kubeClient: kubeClient,
+	}
+}
+
+
+func (c *Controller) Reconcile(ctx context.Context, np *v1beta1.NodePool) (reconcile.Result, error) {
+    nodeClassRef := np.Spec.Template.Spec.NodeClassRef
+    if nodeClassRef == nil {
+        return reconcile.Result{}, fmt.Errorf("nodeClassRef is nil")
+    }
+    // Fetch the resource dynamically using Unstructured
+    config := controllerruntime.GetConfigOrDie()
+    dynamicClient, err := dynamic.NewForConfig(config)
+    if err != nil {
+        return reconcile.Result{}, fmt.Errorf("failed to create dynamic client: %v", err)
+    }
+
+    gvr := schema.GroupVersionResource{
+        Group:    strings.Split(nodeClassRef.APIVersion,"/")[0],
```

Comment:
Rather than parsing like this, we should make use of upstream utility functions. It always feels a bit risky to roll our own logic like this